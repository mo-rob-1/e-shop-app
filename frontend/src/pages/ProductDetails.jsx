import { useState, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { useParams, useNavigate } from "react-router-dom";
import { Link } from "react-router-dom";
import { BsArrowLeft } from "react-icons/bs";

// Actions
import { getProductsDetails } from "../redux/actions/productActions";
import { addToCart } from "../redux/actions/cartActions";

function ProductDetails() {
  const [qty, setQty] = useState(1);

  const dispatch = useDispatch();

  const { id } = useParams();

  const navigate = useNavigate();

  const productDetails = useSelector((state) => state.getProductDetails);
  const { loading, error, product } = productDetails;

  useEffect(() => {
    if (product && id !== product._id) {
      dispatch(getProductsDetails(id));
    }
  }, [dispatch, id, product]);

  const addToCartHandler = () => {
    dispatch(addToCart(product._id, qty));
    navigate(`/cart`);
  };

  return (
    <div className="h-screen">
      {" "}
      {loading ? (
        <h2 className="text-2xl font-bold">Loading...</h2>
      ) : error ? (
        <h2>{error}</h2>
      ) : (
        <>
          <div className="md:mt-8">
            <Link to="/">
              <span className="inline-flex font-semibold items-center shadow pl-5 pr-5 pt-2 pb-2 rounded">
                {" "}
                <BsArrowLeft className="mr-2" />
                Back
              </span>
            </Link>
          </div>
          <div className="mt-8 md:grid md:grid-cols-2 md:gap-10 md:mt-10">
            <div>
              <img src={product.imageUrl} alt={product.name} data-testid="product-details-image" />
            </div>

            <div className="mt-6 md:mt-0">
              <div>
                <div>
                  <h2 className="text-3xl font-bold" data-testid="product-details-name">
                    {product.name}
                  </h2>
                  <p className="mt-4 mb-4 font-bold text-2xl" data-testid="product-details-price">
                    Â£{product.price}
                  </p>

                  <p data-testid="product-details-desc">{product.description}</p>
                </div>
              </div>
              <div>
                <div className="mt-4 grid gap-4">
                  <div>
                    <span className="font-bold mr-2" data-testid="product-details-status">
                      Status:
                    </span>
                    <span>{product.countInStock > 0 ? "In Stock" : "Out of Stock"}</span>
                  </div>
                  <div data-testid="product-details-quantity-selector">
                    <span className="font-bold mr-2" data-testid="product-details-quantity">
                      Quantity:
                    </span>
                    <select className="border" value={qty} onChange={(e) => setQty(e.target.value)}>
                      {[...Array(product.countInStock).keys()].map((x) => (
                        <option key={x + 1} value={x + 1}>
                          {x + 1}
                        </option>
                      ))}
                    </select>
                  </div>

                  <button
                    className="bg-slate-800 text-white text-center pt-3 pb-3 font-semibold mt-3"
                    type="button"
                    onClick={addToCartHandler}
                    data-testid="product-details-add-to-cart-button"
                  >
                    Add To Cart
                  </button>
                </div>
              </div>
            </div>
          </div>
        </>
      )}
    </div>
  );
}

export default ProductDetails;
