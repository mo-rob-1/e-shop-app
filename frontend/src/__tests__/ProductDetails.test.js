import { render } from "@testing-library/react";
import ProductDetails from "../pages/ProductDetails";
import { createMemoryHistory } from "history";
import { Provider } from "react-redux";
import store from "../redux/store";
import { BrowserRouter as Router } from "react-router-dom";

describe("ProductDetails", () => {
  it("renders the product details page", () => {
    const history = createMemoryHistory();
    const { asFragment } = render(
      <Provider store={store} history={history}>
        <Router>
          <ProductDetails />
        </Router>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders the product details page image", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <ProductDetails />
        </Router>
      </Provider>
    );

    const productDetailsImage = getByTestId("product-details-image");
    expect(productDetailsImage).toBeTruthy();
  });

  it("renders the product details page name", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <ProductDetails />
        </Router>
      </Provider>
    );

    const productDetailsName = getByTestId("product-details-name");
    expect(productDetailsName).toBeTruthy();
  });

  it("renders the product details page description", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <ProductDetails />
        </Router>
      </Provider>
    );

    const productDetailsDescription = getByTestId("product-details-desc");
    expect(productDetailsDescription).toBeTruthy();
  });

  it("renders the product details page status", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <ProductDetails />
        </Router>
      </Provider>
    );

    const productDetailsStatus = getByTestId("product-details-status");
    expect(productDetailsStatus).toBeTruthy();
  });

  it("renders the product details page price", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <ProductDetails />
        </Router>
      </Provider>
    );

    const productDetailsPrice = getByTestId("product-details-price");
    expect(productDetailsPrice).toBeTruthy();
  });

  it("renders the product details page quantity", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <ProductDetails />
        </Router>
      </Provider>
    );

    const productDetailsQuantity = getByTestId("product-details-quantity-selector");
    expect(productDetailsQuantity).toBeTruthy();
  });

  it("renders the product details page add to cart button", () => {
    const { getByTestId } = render(
      <Provider store={store}>
        <Router>
          <ProductDetails />
        </Router>
      </Provider>
    );

    const productDetailsAddToCartButton = getByTestId("product-details-add-to-cart-button");
    expect(productDetailsAddToCartButton).toBeTruthy();
  });
});
