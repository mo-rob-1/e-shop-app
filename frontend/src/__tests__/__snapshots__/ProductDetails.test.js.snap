// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductDetails renders the product details page 1`] = `
<DocumentFragment>
  <div>
     
    <div
      class="md:grid md:grid-cols-2 md:gap-10 md:mt-10"
    >
      <div>
        <img
          data-testid="product-details-image"
        />
      </div>
      <div
        class="mt-6 md:mt-0"
      >
        <div>
          <div>
            <h2
              class="text-3xl font-bold"
              data-testid="product-details-name"
            />
            <p
              class="mt-4 mb-4 font-bold text-2xl"
              data-testid="product-details-price"
            >
              Â£
            </p>
            <p
              data-testid="product-details-desc"
            />
          </div>
        </div>
        <div>
          <div
            class="mt-4 grid gap-4"
          >
            <div>
              <span
                class="font-bold mr-2"
                data-testid="product-details-status"
              >
                Status:
              </span>
              <span>
                Out of Stock
              </span>
            </div>
            <div
              data-testid="product-details-quantity-selector"
            >
              <span
                class="font-bold mr-2"
                data-testid="product-details-quantity"
              >
                Quantity:
              </span>
              <select
                class="border"
              >
                <option
                  value="1"
                >
                  1
                </option>
              </select>
            </div>
            <button
              class="bg-slate-800 text-white text-center pt-3 pb-3 font-semibold mt-3"
              data-testid="product-details-add-to-cart-button"
              type="button"
            >
              Add To Cart
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
